tgui/packages/tgui/interfaces/Vending.tsx

```diff
@@ -41,198 +41,210 @@ type HiddenRecord = {
-  const { act, data } = useBackend<VendingData>(context);
-  const {
-    product,
-    productStock,
-    custom,
-  } = props;
+export const Vending = (props, context) => {
+  return (
+    <Window width={450} height={600}>
+      <Window.Content>
+        <Stack fill vertical>
+          <Stack.Item>
+            <UserDetails />
+          </Stack.Item>
+          <Stack.Item grow>
+            <ProductDisplay />
+          </Stack.Item>
+        </Stack>
+      </Window.Content>
+    </Window>
+  );
+};
+
+/** Displays user details if an ID is present and the user is on the station */
+const UserDetails = (_, context) => {
+  const { data } = useBackend<VendingData>(context);
+  const { user, onstation } = data;
+
+  if (!onstation) {
+    return <NoticeBox>Error!</NoticeBox>;
+  } else if (!user) {
+    return (
+      <NoticeBox>No ID detected! Contact the Head of Personnel.</NoticeBox>
+    );
+  } else {
+    return (
+      <Section>
+        <Stack>
+          <Stack.Item>
+            <Icon name="id-card" size={3} mr={1} />
+          </Stack.Item>
+          <Stack.Item>
+            <LabeledList>
+              <LabeledList.Item label="User">{user.name}</LabeledList.Item>
+              <LabeledList.Item label="Occupation">
+                {user.job || 'Unemployed'}
+              </LabeledList.Item>
+            </LabeledList>
+          </Stack.Item>
+        </Stack>
+      </Section>
+    );
+  }
+};
+
+/** Displays  products in a section, with user balance at top */
+const ProductDisplay = (_, context) => {
+  const { data } = useBackend<VendingData>(context);
   const {
-    onstation,
-    department,
     user,
-    jobDiscount,
+    product_records = [],
+    coin_records = [],
+    hidden_records = [],
+    stock,
   } = data;
-  const free = (
-    !onstation
-    || product.price === 0
-    || (
-      !product.premium
-      && department
-      && user
-    )
+  let inventory;
+  let custom = false;
+  if (data.vending_machine_input) {
+    inventory = data.vending_machine_input;
+    custom = true;
+  } else {
+    inventory = [...product_records, ...coin_records];
+    if (data.extended_inventory) {
+      inventory = [...inventory, ...hidden_records];
+    }
+  }
+  // Just in case we still have undefined values in the list
+  inventory = inventory.filter((item) => !!item);
+
+  return (
+    <Section
+      fill
+      scrollable
+      title="Products"
+      buttons={
+        <Box fontSize="16px" color="green">
+          {(user && user.cash) || 0} cr <Icon name="coins" color="gold" />
+        </Box>
+      }>
+      <Table>
+        {inventory.map((product) => (
+          <VendingRow
+            key={product.name}
+            custom={custom}
+            product={product}
+            productStock={stock[product.name]}
+          />
+        ))}
+      </Table>
+    </Section>
   );
+};
+
+/** An individual listing for an item. */
+const VendingRow = (props, context) => {
```

Comment:
This may be me being pedantic, but the VendingRow component drags on vertically for quite some time.

The contents of each table cell could very much be their own component. This time it's not because you're indenting too hard, but because this code is dragging on vertically.

Componentising the contents of the cells will increase code readability and break things up into discrete blocks of behaviour which will be easier to review in isolation.

Reply:
should note that this was the original code. i don't like how it's done, i don't like the usage of tables, and this minor refactor into components will warrant me entirely rewriting the section

Reply:
If you're okay with re-writing it, it would benefit 1000 times over from using our Flex or Flex-derived components.

Reply:
Done. However, I did not aggressively type check the props of the sub components. I can do this if requested